cargo/workspace.bzl

```diff
@@ -0,0 +1,250 @@
+    cargo_label = Label(toolchain_repo + "//:bin/cargo")
+    rustc_label = Label(toolchain_repo + "//:bin/rustc")
+
+    lockfile_path = None
+    if repository_ctx.attr.lockfile:
+        lockfile_path = repository_ctx.path(repository_ctx.attr.lockfile)
+
+    # Yay hand-crafted JSON serialisation...
+    input_content = """{{
+  "repository_name": {name},
+  "packages": [
+      {packages}
+  ],
+  "cargo_toml_files": {{
+      {cargo_toml_files}
+  }},
+  "overrides": {{
+      {overrides}
+  }},
+  "repository_template": "{repository_template}",
+  "target_triples": [
+    {targets}
+  ],
+  "cargo": "{cargo}"
+}}""".format(
+        name = "\"{}\"".format(repository_ctx.attr.name),
+        packages = ",\n".join([artifact for artifact in repository_ctx.attr.packages]),
+        cargo_toml_files = ",\n".join(['"{}": "{}"'.format(ct, repository_ctx.path(ct)) for ct in repository_ctx.attr.cargo_toml_files]),
+        overrides = ",\n".join(["\"{}\": {}".format(oname, ovalue) for (oname, ovalue) in repository_ctx.attr.overrides.items()]),
+        repository_template = repository_ctx.attr.repository_template,
+        targets = ",\n".join(['"{}"'.format(target) for target in repository_ctx.attr.supported_targets]),
+        cargo = repository_ctx.path(cargo_label),
+    )
+
+    input_path = "_{name}.json".format(name = repository_ctx.attr.name)
+
+    # For debugging or working on changes to the resolver, you can set something like:
+    #   export RULES_RUST_RESOLVER_URL_OVERRIDE=file:///path/to/rules_rust/cargo/crate_universe_resolver/target/release/resolver
+    resolver_url = repository_ctx.os.environ.get("RULES_RUST_CRATE_UNIVERSE_RESOLVER_URL_OVERRIDE", None)
+    if resolver_url and resolver_url.startswith("file://"):
+        sha256_result = repository_ctx.execute(["sha256sum", resolver_url[7:]])
+        resolver_sha = sha256_result.stdout[:64]
+
+        resolver_path = repository_ctx.path("resolver")
+        repository_ctx.download(
+            url = resolver_url,
+            sha256 = resolver_sha,
+            output = resolver_path,
+            executable = True,
+        )
+    else:
+        # TODO: Avoid downloading both the linux and darwin one each time.
+        if repository_ctx.os.name == "linux":
+            resolver_path = repository_ctx.path(repository_ctx.attr._resolver_script_linux)
+        elif repository_ctx.os.name == "mac os x":
+            resolver_path = repository_ctx.path(repository_ctx.attr._resolver_script_darwin)
+
+    repository_ctx.file(input_path, content = input_content)
+
+    args = [
+        resolver_path,
+        "--input_path",
+        input_path,
+        "--output_path",
+        repository_ctx.path("defs.bzl"),
+        "--repo-name",
+        repository_ctx.attr.name,
+    ]
+    if lockfile_path != None:
+        args.append("--lockfile")
+        str(args.append(lockfile_path))
+    if repository_ctx.os.environ.get("RULES_RUST_UPDATE_CRATE_UNIVERSE_LOCKFILE", "false") == "true":
```

Comment:
rules_jvm_external just added similar functionality, and they're using `REPIN=1`: https://github.com/bazelbuild/rules_jvm_external/pull/509#discussion_r559893751

https://github.com/bazelbuild/rules_jvm_external/blob/ec2c5617b339844312d4adef4400dcc2ccb73c4f/coursier.bzl#L336-L340

How about we just check that the env var is set (not necessarily to true), and use `REPIN` as well, that way the user can set a single env var to repin everything.

Optionally we could _also_ check for a specific env var like `RULES_RUST_REPIN` so users can opt-in on a per-language basis, but that's also a non-breaking change to add later, so less important.

Reply:
re https://github.com/bazelbuild/rules_rust/pull/598#discussion_r586528064
I feel like it'd make more sense to make updating an explicit run action. How else are users expected to generate a lockfile? Perhaps there's also something I'm missing about this workflow though, can you elaborate a bit?

Reply:
I think I understand the use of `RULES_RUST_UPDATE_CRATE_UNIVERSE_LOCKFILE` now. If the user makes a change to their dependencies, then it won't match the digest in the lockfile and the repository rule will fail. So users will be blocked until they're able to update it by either using this environment variable or commenting out the lockfile. What I still don't understand is how users are expected to extract the lockfile after it's generated. I think there needs to be some run action that will extract and install it, such that users who perform an update would run something like
```
RULES_RUST_REPIN=1 bazel run @my_crate_universe//:repin
```
And this would appropriately update the lockfile related to that repository.